### ISSUE TITLE
Row counts inconsistent

### ISSUE THREAD

**Issue by danluu** [BitFunnel/BitFunnel#275]
I may end up debugging/fixing this as a result of the bug I'm currently chasing, but if not I should go back and figure this out afterwards. This is all using the chunked1 data set from http://bitfunnel.org/wikipedia-as-test-corpus-for-bitfunnel/

There are at least two ways in which row counts are inconsistent. First, `termtable` gives a different row count than `repl status`

Here are the counts from `termtable`:
~~~
===================================
RowAssigner for rank 0
  Terms
    Total: 512640
    Adhoc: 468884
    Explicit: 42357
    Private: 1399

  Rows
    Total: 5814
    Adhoc: 590
    Explicit: 3822
    Private: 1399

...

===================================
RowAssigner for rank 3
  Terms
  Total: 511026                                                                               |            // https://github.com/BitFunnel/BitFunnel/issues/155
    Adhoc: 468884                                                                               |
    Explicit: 34935                                                                             |            size_t adhocRowCount = m_rowAssigners[rank]->GetAdhocRowCount();
    Private: 7207                                                                               |            if (m_termTable.IsRankUsed(rank))
                                                                                                |            {
  Rows                                                                                          |                adhocRowCount = std::max(adhocRowCount, GetMinAdhocRowCount());
    Total: 67599                                                                                |            }
    Adhoc: 24721                                                                                |            m_termTable.SetRowCounts(rank,
    Explicit: 35671                                                                             |                                     m_rowAssigners[rank]->GetExplicitRowCount(),
    Private: 7207
~~~

And here are the counts from `repl status`:
~~~
0: 6228 rows.
1: 1 rows.
2: 1 rows.
3: 67600 rows.
4: 1 rows.
5: 1 rows.
6: 1 rows.
~~~

Second, when I ran `repl status` immediately before running `termtable` again, I got:
~~~
0: 6227 rows.
1: 0 rows.
2: 0 rows.
3: 67599 rows.
4: 0 rows.
5: 0 rows.
6: 0 rows.
~~~

The number of rank0 and rank3 rows increased by 1. I have some debug output statements, but I believe that I haven't changed any code that should affect the actual results. So how come the results of the same command weren't identical?

I haven't fetched since a740476e7955452c2dd6396367c2e8fc341e995d, so I shouldn't have any recent changes that would have affected the number of rows.

When `TermTable::Seal()` is called from `TermTableBuilder`, the rank 0 row counts are as follows:

~~~
m_sharedRowCounts[0] 6224
m_factRowCount 3
m_explicitRowCounts[0] 5224
m_adhocRowCounts[0] 1000
~~~

TermTableBuilder appears to call SetRowCounts with those numbers:

~~~
SetRowCounts
rank 0
explicit 5224
adhoc 1000
SetRowCounts
SetRowCounts
rank 3
explicit 42878
adhoc 24721
~~~

**Comment by danluu** [comment]
I ran the commands a few more times and got the original row counts again:
~~~
0: 6227 rows.
1: 0 rows.
2: 0 rows.
3: 67599 rows.
4: 0 rows.
5: 0 rows.
6: 0 rows.
~~~

The only code changes I made were removing some debug print statements. This seems to indicate that our row mapping is non-deterministic.

**Comment by danluu** [comment]
The mismatched TermTableBuilder / status row count issue has two causes:

1. The fix for #155 made created a local variable that's used to set the ad hoc row count that gets thrown away after being used.

2. Explicit row counts is calculated from `m_bins.size()`, but that's only non-private non-system non-fact explicit rows.

This appears to be unrelated to the non-determinism and also unrelated to the bug where some explicit rows get overlaid on top of the documentActive row.
